client/asset/btc/spv.go

```diff
@@ -0,0 +1,1601 @@
+	// by whether blockHash is nil, since that means it wasn't passed in, and
+	// it wasn't found in the database.
+	if blockHash == nil {
+		var err error
+		_, limitHeight, err = w.findBlockForTime(startTime)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Do a filter scan.
+	utxo, err := w.filterScanFromHeight(*txHash, vout, pkScript, limitHeight, checkPt)
+	if err != nil {
+		return nil, fmt.Errorf("filterScanFromHeight error: %w", err)
+	}
+	if utxo == nil {
+		return nil, asset.CoinNotFoundError
+	}
+
+	// If we found a block, let's store a reference in our local database so we
+	// can maybe bypass a long search next time.
+	if utxo.blockHash != nil {
+		w.storeTxBlock(*txHash, *utxo.blockHash)
+	}
+
+	w.cacheCheckpoint(txHash, vout, utxo)
+
+	return utxo, nil
+}
+
+// getTxOut finds an unspent transaction output and its number of confirmations.
+// To match the behavior of the RPC method, even if an output is found, if it's
+// known to be spent, no *wire.TxOut and no error will be returned.
+func (w *spvWallet) getTxOut(txHash *chainhash.Hash, vout uint32, pkScript []byte, startTime time.Time) (*wire.TxOut, uint32, error) {
+	// Check for a wallet transaction first
+	txDetails, err := w.wallet.walletTransaction(txHash)
+	if err == nil {
+		tip, err := w.cl.BestBlock()
+		if err != nil {
+			return nil, 0, fmt.Errorf("BestBlock error: %v", err)
+		}
+
+		var confs uint32
+		if txDetails.Block.Height > 0 {
+			confs = uint32(tip.Height - txDetails.Block.Height + 1)
+		}
+		msgTx := &txDetails.MsgTx
+		if len(msgTx.TxOut) < int(vout+1) {
+			return nil, 0, fmt.Errorf("wallet transaction %s found, but not enough outputs for vout %d", txHash, vout)
+		}
+		return msgTx.TxOut[vout], confs, nil
+	}
+	if !errors.Is(err, WalletTransactionNotFound) {
+		return nil, 0, fmt.Errorf("walletTransaction error: %w", err)
+	}
+
+	var blockHash *chainhash.Hash
+	if txDetails != nil && txDetails.Block.Hash != (chainhash.Hash{}) {
+		blockHash = &txDetails.Block.Hash
+	}
+
+	// We don't really know if it's spent, so we'll need to scan.
+	utxo, err := w.scanFilters(txHash, vout, pkScript, startTime, blockHash)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if utxo == nil || utxo.spend != nil {
+		return nil, 0, nil
+	}
+
+	if utxo.blockHash == nil {
+		return nil, 0, fmt.Errorf("output %s:%v not found", txHash, vout)
+	}
+
+	tip, err := w.cl.BestBlock()
+	if err != nil {
+		return nil, 0, fmt.Errorf("BestBlock error: %v", err)
+	}
+
+	confs := uint32(tip.Height) - utxo.blockHeight + 1
+
+	return utxo.txOut, confs, nil
+}
+
+// filterScanFromHeight scans BIP158 filters beginning at the specified block
+// height until the tip, or until a spending transaction is found.
+func (w *spvWallet) filterScanFromHeight(txHash chainhash.Hash, vout uint32, pkScript []byte, startBlockHeight int32, checkPt *filterScanResult) (*filterScanResult, error) {
+	tip, err := w.getBestBlockHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	res := checkPt
```

Comment:
I'm thinking we want to make a copy of the struct itself because we modify fields below, and this `*filterScanResult` is still in the checkpoints map and thus accessible by other goroutines.

Although I don't yet have a sense for if it is possible for simultaneous `scanFilters` calls for the same output.